src/main/java/com/google/sps/servlets/TagServlet.java

```diff
@@ -0,0 +1,82 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.meltingpot.servlets;
+
+import com.google.api.core.ApiFuture;
+import com.google.cloud.firestore.CollectionReference;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.Query;
+import com.google.cloud.firestore.QueryDocumentSnapshot;
+import com.google.cloud.firestore.QuerySnapshot;
+import com.google.cloud.firestore.WriteResult;
+import com.google.firebase.auth.FirebaseToken;
+import com.google.firebase.cloud.FirestoreClient;
+import com.google.gson.Gson;
+import com.google.gson.reflect.TypeToken;
+import com.google.sps.meltingpot.auth.Auth;
+import com.google.sps.meltingpot.data.DBInterface;
+import com.google.sps.meltingpot.data.DBObject;
+import com.google.sps.meltingpot.data.FirestoreDB;
+import com.google.sps.meltingpot.data.Recipe;
+import com.google.sps.meltingpot.data.Tag;
+import com.google.sps.meltingpot.data.User;
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+import javax.servlet.AsyncContext;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Handles recipe post requests. */
+@WebServlet("/api/tag")
+public class TagServlet extends HttpServlet {
+  private Gson gson = new Gson();
+  private DBInterface db;
+
+  public TagServlet(DBInterface db) {
+    this.db = db;
+  }
+
+  @Override
+  public void init() {
+    db = new FirestoreDB();
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String[] tagIds = request.getParameterValues("tagIds");
+    String getHidden = request.getParameter("getHidden");
+
+    List<Tag> tags;
+
+    if (tagIds == null) {
```

Comment:
what happens if tagIds exist but has no value? like:
https://our.url.com/Tagservlet?tagIds=&getHidden=true?

Reply:
That would count as the empty string, and since no tag has the empty string for an id, it would simply return an empty array.